packages/asset-swapper/src/network/samplers/balancer.ts

```diff
@@ -0,0 +1,250 @@
+                    } catch {
+                        // soldier on
+                    }
+                }
+            }
+        }
+    }
+
+    protected async _fetchTopPoolsAsync(): Promise<BalancerPoolResponse[]> {
+        const query = gql`
+            query fetchTopPools($topPoolsFetched: Int!) {
+                pools(
+                    first: $topPoolsFetched
+                    where: { publicSwap: true, liquidity_gt: 0 }
+                    orderBy: swapsCount
+                    orderDirection: desc
+                ) {
+                    id
+                    publicSwap
+                    swapFee
+                    totalWeight
+                    tokensList
+                    tokens {
+                        id
+                        address
+                        balance
+                        decimals
+                        symbol
+                        denormWeight
+                    }
+                }
+            }
+        `;
+        try {
+            const { pools } = await request(this._subgraphUrl, query, { topPoolsFetched: this._topPoolsFetched });
+            return pools;
+        } catch (err) {
+            return [];
+        }
+    }
+}
+
+export class CreamPoolsCache extends PoolsCache {
+    constructor(
+        _cache: { [key: string]: CacheValue } = {},
+        private readonly maxPoolsFetched: number = BALANCER_MAX_POOLS_FETCHED,
+    ) {
+        super(_cache);
+    }
+
+    protected async _fetchPoolsForPairAsync(takerToken: string, makerToken: string): Promise<Pool[]> {
+        try {
+            const poolData = (await getCreamPoolsWithTokens(takerToken, makerToken)).pools;
+            // Sort by maker token balance (descending)
+            const pools = parseCreamPoolData(poolData, takerToken, makerToken).sort((a, b) =>
+                b.balanceOut.minus(a.balanceOut).toNumber(),
+            );
+            return pools.slice(0, this.maxPoolsFetched);
+        } catch (err) {
+            return [];
+        }
+    }
+}
+
+type SellContract = ERC20BridgeSamplerContract;
+type BuyContract = ERC20BridgeSamplerContract;
+type SellContractSellFunction = SellContract['sampleSellsFromBalancer'];
+type BuyContractBuyFunction = BuyContract['sampleBuysFromBalancer'];
+type SellContractSellFunctionParameters = Parameters<SellContractSellFunction>;
+type BuyContractBuyFunctionParameters = Parameters<BuyContractBuyFunction>;
+type SellContractSellFunctionReturn = UnwrapContractFunctionReturnType<ReturnType<SellContractSellFunction>>;
+type BuyContractBuyFunctionReturn = UnwrapContractFunctionReturnType<ReturnType<BuyContractBuyFunction>>;
+type SamplerSellEthCall = SamplerEthCall<BalancerFillData,SellContractSellFunctionParameters,SellContractSellFunctionReturn>;
+type SamplerBuyEthCall = SamplerEthCall<BalancerFillData,BuyContractBuyFunctionParameters,BuyContractBuyFunctionReturn>;
```

Comment:
This is bit of unfortunately required boilerplate for each source to make the rest of the file more legible and nudge generic type inference along.

Reply:
It's pretty easy to copy & paste & update for new sources though.